text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+[def-package]:           https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package           '"package" (glossary entry)'
+[def-package-registry]:  https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package-registry  '"package-registry" (glossary entry)'
+[def-manifest]:          https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#manifest          '"manifest" (glossary entry)'
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Single-file packages
+
+In addition to today's multi-file packages (`Cargo.toml` file with other `.rs`
+files),
+we are adding the concept of single-file packages which may contain an
+embedded manifest.
+There is no required way to distinguish a single-file
+`.rs` package from any other `.rs` file.
+
+A single-file package may contain an embedded manifest.
+An embedded manifest is stored using `TOML` in rust "frontmatter",
+a markdown code-fence with `cargo` at the start of the infostring at the top of
+the file.
+
+Inferred / defaulted manifest fields:
+- `package.name = <slugified file stem>`
+- `package.version = "0.0.0"` to [call attention to this crate being used in unexpected places](https://matklad.github.io/2021/08/22/large-rust-workspaces.html#Smaller-Tips)
+  - This may be altered with [rust-lang/cargo#9829](https://github.com/rust-lang/cargo/issues/9829)
+- `package.publish = false` to avoid accidental publishes, particularly if we
+  later add support for including them in a workspace.
+- `package.edition = <current>` to avoid always having to add an embedded
+  manifest at the cost of potentially breaking scripts on rust upgrades
+  - Warn when `edition` is unspecified.  Since piped commands are run with `--quiet`, this may not show up.
+  - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand
+  - Based on feedback, we can switch to "edition is required as of <future> edition"
+
+Disallowed manifest fields:
+- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
+- `package.workspace`, `package.build`, `package.links`, `package.publish`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`
+
+Single-file packages maintain an out-of-line target directory by default.
+This is `$CARGO_HOME/target/<hash-of-path>`.
+
+A single-file package is accepted by cargo commands as a `--manifest-path`
+- Files are considered to have embedded manifest if they end with `.rs` or they lack and extension and are of type file
+- This allows running `cargo test --manifest-path single.rs`
+- `cargo add` and `cargo remove` may not support editing embedded manifests initially
+- Path-dependencies may not refer to single-file packages at this time (they don't have a `lib` target anyways)
+
+Single-file packages will not be accepted as `path` or `git` dependencies.
+
+The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
```

Comment:
This means that lockfiles, as well as build artefacts, will just be placed in a `target` dir next to the script, right? I think this is one of the most problematic parts of this proposals. Potential problems:

- People won't know that there's a lockfile and keep using very old versions of security-relevant dependencies
- Running two scripts in the same directory via Cargo will lead to one's lockfile overwriting the other one's (?)
- It seems annoying to have a visible `target` directory at all

I will leave my other thoughts on this in the future possibilities section.